### PR TITLE
planner: add optimizer trace framework for logicalOptimize

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Add optimize trace framework for logicalOptimize. In this pr, we do the following work.

1. use stmtCtx.EnableOptimizeTrace to indicate whether the statement should be enabled optimizer trace
2. add optimize trace common framework into logical optimize processing.

Issue Number: close #xxx

Problem Summary:

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/optimizer.go [hunk 1]
```diff
@@ -83,9 +84,25 @@ var optRuleList = []logicalOptRule{
...
+type logicalOptimizeOp struct {
+	enableTrace bool
+}
+
+type logicalOptimizeOption func(*logicalOptimizeOp)
+
+func withEnableOptimizeTrace() logicalOptimizeOption {
+	return func(op *logicalOptimizeOp) { op.enableTrace = true }
+}
+
```
Comment (rebelice) [inline planner/core/optimizer.go:96]: Why do we need this function, it does not seem to be necessary at the moment. What will be added here later?

Comment (Yisaer) [reply planner/core/optimizer.go:96]: `withEnableOptimizeTrace ` is used in `rule.optimize`. This lets each rule knows whether it needs to collect trace information.

Comment (rebelice) [reply planner/core/optimizer.go:96]: It seems that there is not much difference between using `logicalOptimizeOp`. why do we use function parameters?

File: planner/core/rule_aggregation_elimination.go [hunk 2]
```diff
@@ -179,25 +179,25 @@ func wrapCastFunction(ctx sessionctx.Context, arg expression.Expression, targetT
 	return expression.BuildCastFunction(ctx, arg, targetTp)
 }
 
-func (a *aggregationEliminator) optimize(ctx context.Context, p LogicalPlan) (LogicalPlan, error) {
+func (a *aggregationEliminator) optimize(ctx context.Context, p LogicalPlan, opts ...logicalOptimizeOption) (*logicalOptimizeReturn, error) {
 	newChildren := make([]LogicalPlan, 0, len(p.Children()))
 	for _, child := range p.Children() {
 		newChild, err := a.optimize(ctx, child)
```
Comment (rebelice) [inline planner/core/rule_aggregation_elimination.go:185]: Do we need to add `opts` here?

File: planner/core/rule_decorrelate.go [hunk 3]
```diff
@@ -161,8 +161,8 @@ func (s *decorrelateSolver) optimize(ctx context.Context, p LogicalPlan) (Logica
 				if err != nil {
 					return nil, err
 				}
-				proj.SetChildren(np)
-				return proj, nil
+				proj.SetChildren(np.logicalPlan)
+				return &logicalOptimizeReturn{logicalPlan: proj}, nil
 			}
 			return s.optimize(ctx, p)
```
Comment (rebelice) [inline planner/core/rule_decorrelate.go:167]: ditto

**Review by rebelice — COMMENTED** [review]
rest LGTM

File: planner/core/rule_aggregation_elimination.go (hunk 2 above)
Comment (Yisaer) [reply planner/core/rule_aggregation_elimination.go:185]: currently we only implement framework for rule.optimize. In future we will implement specific logical code for each rule during optimize. Thus no need to add `opts` here for now.

File: planner/core/optimizer.go (hunk 1 above)
Comment (Yisaer) [reply planner/core/optimizer.go:96]: We are using [functional options](https://www.sohamkamani.com/golang/options-pattern/) here

**Review by rebelice — APPROVED** [review]

File: planner/core/plan.go [hunk 4]
```diff
@@ -377,6 +381,15 @@ func (p *baseLogicalPlan) ExplainInfo() string {
 	return ""
 }
 
+// buildLogicalPlanTrace implements LogicalPlan
+func (p *baseLogicalPlan) buildLogicalPlanTrace() tracing.LogicalPlanTrace {
+	planTrace := tracing.LogicalPlanTrace{ID: p.ID(), TP: p.TP()}
```
Comment (AilinKid) [inline planner/core/plan.go:386]: explainInfo will be filled next?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: planner/core/plan.go [hunk 5]
```diff
@@ -377,6 +381,15 @@ func (p *baseLogicalPlan) ExplainInfo() string {
 	return ""
 }
 
+// buildLogicalPlanTrace implements LogicalPlan
+func (p *baseLogicalPlan) buildLogicalPlanTrace() tracing.LogicalPlanTrace {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply planner/core/plan.go]: updated.

File: planner/core/plan.go (hunk 4 above)
Comment (Yisaer) [reply planner/core/plan.go]: yes, `explainInfo` will be implemented by specific logical operator.

**Review by AilinKid — APPROVED** [review]

File: planner/core/optimizer.go [hunk 6]
```diff
@@ -83,9 +84,25 @@ var optRuleList = []logicalOptRule{
 	&columnPruner{}, // column pruning again at last, note it will mess up the results of buildKeySolver
 }
 
+type logicalOptimizeOp struct {
+	enableTrace bool
```
Comment (chrysan) [inline planner/core/optimizer.go:88]: If we are not going to have more options here in the foreseeable future, we can simplify.

File: planner/core/optimizer.go [hunk 7]
```diff
@@ -343,11 +364,43 @@ func logicalOptimize(ctx context.Context, flag uint64, logic LogicalPlan) (Logic
...
+func logicalOptimizeWithTraceEnabled(ctx context.Context, flag uint64, logic LogicalPlan) (LogicalPlan, error) {
+	trace := &tracing.LogicalOptimizeTrace{
+		Steps: make(map[int]tracing.LogicalRuleOptimizeTrace),
+	}
+	var err error
+	for i, rule := range optRuleList {
+		// The order of flags is same as the order of optRule in the list.
+		// We use a bitmask to record which opt rules should be used. If the i-th bit is 1, it means we should
+		// apply i-th optimizing rule.
+		if flag&(1<<uint(i)) == 0 || isLogicalRuleDisabled(rule) {
```
Comment (chrysan) [inline planner/core/optimizer.go:385]: This part is duplicated with the original one, you can align. Please think about minimizing changes to original codes to make tracer work.

File: planner/core/optimizer.go [hunk 8]
```diff
@@ -83,9 +84,25 @@ var optRuleList = []logicalOptRule{
...
+type logicalOptimizeReturn struct {
+	logicalPlan LogicalPlan
+	// comment it for now in order to pass lint, we will use steps in future
+	// optimizeTraceSteps []tracing.LogicalRuleOptimizeTraceStep
+}
+
 // logicalOptRule means a logical optimizing rule, which contains decorrelate, ppd, column pruning, etc.
 type logicalOptRule interface {
-	optimize(context.Context, LogicalPlan) (LogicalPlan, error)
+	optimize(context.Context, LogicalPlan, ...logicalOptimizeOption) (*logicalOptimizeReturn, error)
```
Comment (chrysan) [inline planner/core/optimizer.go:105]: Please add one rule implementation to show how this framework works.

**Review by chrysan — CHANGES_REQUESTED** [review]

File: planner/core/optimizer.go (hunk 7 above)
Comment (Yisaer) [reply planner/core/optimizer.go:381]: updated.

File: planner/core/optimizer.go (hunk 8 above)
Comment (Yisaer) [reply planner/core/optimizer.go]: updated in `aggregationEliminateChecker`.

File: planner/core/rule_aggregation_elimination.go [hunk 9]
```diff
@@ -100,6 +101,16 @@ func (a *aggregationEliminateChecker) tryToEliminateDistinct(agg *LogicalAggrega
 				}
 				if distinctByUniqueKey {
 					af.HasDistinct = false
+					if opt.tracer != nil {
+						opt.tracer.AppendRuleTracerStepToCurrent(
+							tracing.LogicalRuleOptimizeTraceStep{
```
Comment (chrysan) [inline planner/core/rule_aggregation_elimination.go:106]: Do not put trace details in rule logic, move them to a tracer func, just pass planNode/expression... Each struct/function should have clean responsibility.

Comment (Yisaer) [reply planner/core/rule_aggregation_elimination.go]: updated.

File: planner/core/optimizer.go [hunk 10]
```diff
@@ -83,9 +84,44 @@ var optRuleList = []logicalOptRule{
...
+func defaultLogicalOptimizeOption() *logicalOptimizeOp {
+	return &logicalOptimizeOp{}
+}
+
+func (op *logicalOptimizeOp) withEnableOptimizeTrace(tracer *tracing.LogicalOptimizeTracer) *logicalOptimizeOp {
+	op.tracer = tracer
+	return op
+}
+
+func (op *logicalOptimizeOp) appendRuleTracerBeforeRuleOptimize(name string, before LogicalPlan) {
```
Comment (chrysan) [inline planner/core/optimizer.go:101]: If you want logicalOptimizeOp to be generic not only for tracer, you'd better set the func name more generic without word "tracer".

Comment (Yisaer) [reply planner/core/optimizer.go]: updated.

**Pull request merged**